Support correct `$crate` expansion in macros

uHOOCCOOHu opened the pull request:
This PR makes normal use cases of `$crate` from macros work as expected.

It makes more macros from `std` work. Type inference works well with `panic`, `unimplemented`, `format`, and maybe more.
Sadly that `vec![1, 2, 3]` still not works, but it is not longer an issue about macro.

Screenshot:
![Screenshot_20190927_022136](https://user-images.githubusercontent.com/14816024/65714465-b4568f80-e0cd-11e9-8043-dd44c2ae8040.png)

uHOOCCOOHu commented:
@matklad
> ..., but, when we lower ast to paths, we make sure to convert $crate to a Repr::DollarCrate with the appropriate crate id.

Do you mean we need to pass `CrateId` through `parse_or_expand` [here](https://github.com/rust-analyzer/rust-analyzer/pull/1815/files#diff-51cf0d02787e805fff19a8238d651d8eR464)? But all `MacroCall` must do this replacement and it will become kind of redundant.

Well, in fact, I don't know how to get this information in `ast::Name::as_name` without passing it through `SyntaxToken::text`.

matklad commented:
In general, it seems like any call to `Path::from_ast` should gain an additional parameter, the file where the pat resides, such that `Path::from_ast` can figure out the meaning of the `$crate` if the file comes from macro expansion

uHOOCCOOHu commented:
> In general, it seems like any call to Path::from_ast should gain an additional parameter, the file where the pat resides, such that Path::from_ast can figure out the meaning of the $crate if the file comes from macro expansion

Oops, since `Path::from_ast` is called from macro caller's context, the `current_file_id` will not be the crate containing `macro_rules` if it's a external macro.
If we lose this information in AST, we could never recover it in higher level.

matklad commented:
`current_file_id` is the `HirFileId`. For macro files, it is `MacroFile`, which contains a `MacroCallId`, which contains `MacroDefId`, and the crate of this `MacroDefId` is the crate we need. 

That is, we identify macro calls by *both* the location of the call to the macro and the definition of the macro, and we can fetch the correct crate from the definition.

matklad commented:
(we'd still need to add a `Crate` to `MacroDefId`)

uHOOCCOOHu commented:
@matklad
I found in rustc, expansion of `$crate` is a single `tt`,  `stringify!` it results in `"$crate"` (no space inside), and passing it through macros to construct `use crate::$crate` will result in `no '$crate' in the root`.
It seems weird but just behaves the same as your design. :laughing:

matklad commented:
Note that for hygiene we'll have to augment AST itself with additional info, but I hope not to embedd the info direclty into the AST, and instead store it on the side.

Specifically, mbe tarnsform TokenTrees, and tokens in tts have identity. So, it is possible to correlate input tokens with output tokens. When we parse the token stream into AST, we should take care to "color" each identifier as comming from the macro argument, or from the macro expansion, using this info. But we'll store colors alongside the tree, as `FxHashMap<TextRange, ExpansionId>`.

uHOOCCOOHu commented:
> `current_file_id` is the `HirFileId`. For macro files, it is `MacroFile`, which contains a `MacroCallId`, which contains `MacroDefId`, and the crate of this `MacroDefId` is the crate we need.
> 
> That is, we identify macro calls by _both_ the location of the call to the macro and the definition of the macro, and we can fetch the correct crate from the definition.

I'm trying it, but seems it cause tons of modification. Including `Path::from_ast`, `Generic::{from_ast,fill_params}`, some methods in `RawItemsCollector`, `SourceAnalyzer` and `TypeRef`, and more.
I'm curios that if it is the right way to do this. The code to parse paths are everywhere. 

For hygiene, note that identifiers are colored by the macro call id. And the color of identifier is only used in expression or statement place. Items defined by macros can always escape.

Will it be easier to use hygiene for `$crate`?

matklad commented:
I believe that, to support hygiene, there's no way around changing every call-site where we convert AST to hir. An isolated `ast::Path` is fundamentally a lossy type, which doesn't have affordance for hygiene. 

But I'll be fine with doing this incrementally: instead of changing `from_ast`, introduce a `from_ast_with_hygiene` method alongside, and leisurely remove usages of the former, polishing the API along the way.

However, I think we've recently distilled a good abstraction to help with is: `ast::Whatever` is a value-type ast. Two textually equal whatevers in two different files can be represented by a single ast node (ast is ref-counted inside). And `Source<ast::Whatever>` is a specific *instance* of `ast::Whatever`, which is an AstValue plus associated identity (HirFileId). So, I think we need

```rust
impl Path {
    fn from_src(src: Source<ast::Path>)
}
```

>Will it be easier to use hygiene for $crate?

I think now, as I imagine hyginee looking as `DB::hygiene_info(file_id: HirFileId) -> FxHashMap<TextRange, HirFIleId>`. Ie, you still need to thread `HirFileId`, but you have a new db query on top!.

matklad commented:
@uHOOCCOOHu do you plan to follow up here? I am extremely excited about this, as figuring out hygiene is one of the few yet unresolved questions in rust-analyzer's architecture!

uHOOCCOOHu commented:
@matklad 
Oh, just kind of busy recently. I'll look into this later this week.

I'm not planning to support full featured hygienic macro here,  though `$crate` can be a part of it.

uHOOCCOOHu commented:
Refactored and rebased.

Now it produces identifier `$crate` when expanding macros, and it gets resolved in `Path::from_src` into a special `PathKind::DollarCrate`.
Macros in item and expression place are supported. Type place macros are marked as `FIXME` since it need more modification.

I keep the refactored version of `Name`. It would help when implementing hygienic macro (not now). But `$crate` is handled outside `Name`, so it is not used currently.

matklad commented on crates/ra_hir/src/nameres.rs:
```diff
@@ -332,6 +332,20 @@ impl CrateDefMap {
+                            tested_by!(macro_dollar_crate_other);
+                            PerNs::types(module.into())
+                        }
+                        None => return ResolvePathResult::empty(ReachedFixedPoint::No),
```
```suggestion
                        None => return ResolvePathResult::empty(ReachedFixedPoint::Yes),
```

There's no way this path can be resolved later, so empty resolution is a fixed point here

matklad commented on crates/ra_hir/src/nameres/raw.rs:
```diff
@@ -192,13 +194,15 @@ pub(super) struct MacroData {
 }
 
-impl RawItemsCollector {
+impl<DB: AstDatabase> RawItemsCollector<&'_ DB> {
```
```suggestion
impl<DB: AstDatabase> RawItemsCollector<&DB> {
```

uHOOCCOOHu commented:
Fixed.

bors r+

lnicola commented:
```before
Database loaded, 214 roots, 227.204874ms
Crates in this dir: 25
Total modules found: 298
Total declarations: 10291
Total functions: 3622
Item Collection: 13.806968216s, 0b allocated 0b resident
Total expressions: 81672
Expressions of unknown type: 9007 (11%)
Expressions of partially unknown type: 4205 (5%)
Type mismatches: 2774
Inference: 31.848437435s, 0b allocated 0b resident
Total: 45.655412334s, 0b allocated 0b resident

after
Database loaded, 214 roots, 260.188944ms
Crates in this dir: 25
Total modules found: 299
Total declarations: 10342
Total functions: 3649
Item Collection: 11.624890941s, 0b allocated 0b resident
Total expressions: 82923
Expressions of unknown type: 6006 (7%)
Expressions of partially unknown type: 3307 (3%)
Type mismatches: 4035
Inference: 24.926154897s, 0b allocated 0b resident
Total: 36.551052608s, 0b allocated 0b resident
```
